validator/index.js

```diff
@@ -25,50 +25,233 @@ const path = require('path');
+     * context for the error. In a user interface this would be shown
+     * as a "Learn more" link.
+     * @type {!string}
+     */
+    this.specUrl = null;
+    /**
+     * Categorizes error messages into higher-level groups. This makes it
+     * easier to create error statistics across a site and give advice based
+     * on the most common problems for a set of pages.
+     * See the ErrorCategory.Code enum in validator.proto for possible values.
+     * @type {!string}
+     */
+    this.category = 'UNKNOWN';
+    /**
+     * This field is only useful when scripting against the validator,
+     * it should not be displayed in a user interface as it adds nothing
+     * for humans to read over the message field (see above).
+     * Possible values are the codes listed in ValidationError.Code - see
+     * validator.proto. Examples: 'UNKNOWN_CODE', 'MANDATORY_TAG_MISSING',
+     * 'TAG_REQUIRED_BY_MISSING'. For each of these codes there is a
+     * format string in validator-main.protoascii (look for error_formats),
+     * which is used to assemble the message from the strings in params.
+     * @type {!string}
+     */
+    this.code = 'UNKNOWN_CODE';
+    /**
+     * This field is only useful when scripting against the validator,
+     * it should not be displayed in a user interface as it adds nothing
+     * for humans to read over the message field (see above).
+     * @type {!Array<!string>}
+     */
+    this.params = [];
+  }
+}
+
+/**
+ * A global static map used by the validateString function to avoid loading
+ * AMP Validators more than once.
+ * @type {!Object<string, Object>}
+ */
+const precompiledByValidatorJs = {};
+
+/**
+ * @param {!string} inputString
+ * @param {string=} validatorJs_opt
+ * @returns {Promise<(ValidationResult|Error)>}
+ * @export
+ */
+function validateString(inputString, opt_validatorJs) {
+  const validatorJs =
+      opt_validatorJs || 'https://cdn.ampproject.org/v0/validator.js';
+  // The 'sandbox' is a Javascript object (dictionary) which holds the results
+  // of evaluating the validatorJs, so basically, it holds functions, prototypes,
+  // etc. As a side-effect of evaluating, the VM will compile this code and it's
+  // worth holding onto it. Hence, this validate function is reached via
+  // 2 codepaths - either the sandbox came from the cache,
+  // precompiledByValidatorJs - or we just constructed it after downloading
+  // and evaluating the script. The API is fancier here, vm.Script /
+  // vm.createContext / vm.runInContext and all that, but it's quite similar
+  // to a Javascript eval.
+  const validate = (sandbox, resolve) => {
+    const internalResult = sandbox.amp.validator.validateString(inputString);
+    const result = new ValidationResult();
+    result.status = internalResult.status;
+    for (const internalError of internalResult.errors) {
+      const error = new ValidationError();
+      error.severity = internalError.severity;
```

Comment:
This is neat, but I really want to keep this explicit. I need to do a little bit more work there than just copying parameters, since we're switching the enums to use numbers instead of strings, but I still want to use strings for this API.